OSDOCS-6248:adds AdminNetworkPolicy to OVN-K (#70800)
openshift/openshift-docs: modules/nw-ovn-k-baseline-adminnetwork-policy.adoc

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: <Concept>
[id="BaselineAdminNetworkPolicy"_{context}"]
= BaselineAdminNetworkPolicy

> `BaselineAdminNetworkPolicy` (BANP) is a cluster-scoped custom resource definitions (CRDs). As an {product-title} administrator, you can use `BaselineAdminNetworkPolicy` to setup and enforce optional baseline network policies that are overridable by users if need be. Rule actions for BANP are `allow` or `deny`.

Reviewer: ```suggestion
`BaselineAdminNetworkPolicy` (BANP) is a cluster-scoped custom resource definition (CRD) shipped and maintained by blah -> same as suggested for ANP. As an {product-title} administrator, you can use `BaselineAdminNetworkPolicy` to setup and enforce optional baseline network policies that are overridable by users using NetworkPolicies if need be.
```

Rule actions for BANP are `allow` or `deny` -> we probably want to put this somewhere else like near to the BANP example action line? (I think you already have...)

BANP allows administrators to specify:

* A `subject` that consists of a set of pods that is selected by the policy.

* A list of ingress rules to be applied for all ingress traffic towards the `subject`.

* A list of egress rules to be applied for all egress traffic from the `subject`.

[discrete]
[id="BaselineAdminNetworkPolicy-example"]
== BaselineAdminNetworkPolicy example

> The `BaselineAdminNetworkPolicy` resource is a cluster singleton `object` that can be used as a guardrail policy incase a passed traffic policy does not match any `NetworkPolicies` in the cluster. You must use `default` as the name when creating a `BaselineAdminNetworkPolicy` object.

Reviewer: It can also be used if there are no ANPs and NPs at all... so not just for PASS, it can also be just used as is for default guardrails use case (please the main KEP I had linked in slack for more details)

> // Need to confirm with SME if this is correct`BaselineAdminNetworkPolicy` is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. See "Enabling features using feature gates" in the Additional resources of this section for more information on feature gates and `TechnologyPreviewNoUpgrade` features.

Reviewer: So BANP is available along with ANP actually with the same flag.

So the feature is called "AdminNetworkPolicy" but has two APIs/CRDs called ANP and BANP. I hope that makes sense...

.Example YAML file for BANP
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:

>   name: priority-50-example <1>

Reviewer: bad name, there is no priority for BANP
OH WAIT, this is ANP, you need to change this to a BANP..

spec:
  priority: 50 <2>
  subject:
    namespaces:
      matchLabels:
          kubernetes.io/metadata.name: gryffindor
  ingress:<3>
  - name: "deny-all-ingress-from-slytherin"
    action: "Deny"
    from:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              conformance-house: slytherin
        podSelector:
          matchLabels:
            conformance-house: slytherin
  egress:<3>
  - name: "pass-all-egress-to-slytherin"
    action: "Allow" <4>
    to:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              conformance-house: slytherin
        podSelector:
          matchLabels:
            conformance-house: slytherin
----
<1> Policy name

> <2> `spec.priority` supports a maximum of 100 BANP in the values of 0-99 in a cluster. The lower the value the higher precedence. Creating `BaselineAdminNetworkPolicy` with the same priority because the outcome is nondeterministic.

Reviewer: no priorities

<3> BANP have both ingress and egress rules.
<4> BANP `action` rules for ingress and egress accepts values of `Deny` and `Allow`.
====

[discrete]
[id="BaselineAdminNetworkPolicy-default-deny-example"_{context}]
== BaselineAdminNetworkPolicy Deny example
The following BANP singleton ensures that the administrator has set up a default deny policy for all ingress monitoring traffic coming into the tenants at `internal` security level. When combined with the "AdminNetworkPolicy Pass example", this deny policy acts as a guardrail policy for all ingress traffic that is passed by the ANP `pass-monitoring` policy.

.Example YAML file for a guardrail Deny rule
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces:
      matchLabels:
        security: internal
  ingress:
  - name: "deny-ingress-from-monitoring"
    action: "Deny"
    from:
    - namespaces:
        namespaceSelector:
          matchLabels:
# ...
----
====

[discrete]
[id="BaselineAdminNetworkPolicy-multi-tenancy-example"_{context}]
== BaselineAdminNetworkPolicy multi-tenancy example

> The following example demonstrates how `AdminNetworkPolicy` with `Pass` action can be used by administrators in conjunction with `BaseleineNetworkPolicy to create a multi-tenancy policy that allows one tenant to collect monitoring data on their application while simultaneously not collecting data from the second tenant.

Reviewer: Something is wrong here... ? ANP/NP/BANP are missing or are you still writing this section? I am not sure I follow the pieces here..

Author: I honestly think this section should be pulled out till [4.16](https://issues.redhat.com/browse/OSDOCS-9439) and added into the common use cases modules.

> As an administrator if you apply both the "AdminNetworkPolicy Pass action example" and the "BaselineNetwork Policy Deny example", tenants are then left with the ability to choose to create a `NetworkPolicy` that will be evaluated before the BANP.

Author: I structured this multi-tenancy based on your blogpost which is where I took the examples from for ANP Pass action and BANP Deny. When you combine those examples with this, you get the multi-tenancy policy I'm discussing here and you were discussing in your blog post. @tssurya. Which is to say, this is not clear and needs to come out for not till 4.16, I think.

Reviewer: but when I took an initial review stab at this it was totally incomplete, I think the pieces are now complete based on the blogpost and make more sense. They can stay and are pretty useful and should tie up loose ends with the diagram from top. Having said that I can't review these bits because I wrote them so refrainng from doing so..

<4> Specify a name for the `ingress.name`
<5> Specify the namespaces to select the pods from to apply the BANP resource.
<6> Specify `podSelector.matchLabels` name of the pods to apply the BANP resource.
====

[discrete]
[id="BaselineAdminNetworkPolicy-default-deny-example"_{context}]
== BaselineAdminNetworkPolicy Deny example
The following BANP singleton ensures that the administrator has set up a default deny policy for all ingress monitoring traffic coming into the tenants at `internal` security level. When combined with the "AdminNetworkPolicy Pass example", this deny policy acts as a guardrail policy for all ingress traffic that is passed by the ANP `pass-monitoring` policy.

.Example YAML file for a guardrail `Deny` rule
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces:
      matchLabels:
        security: internal
  ingress:
  - name: "deny-ingress-from-monitoring"
    action: "Deny"
    from:
    - namespaces:
        namespaceSelector:
          matchLabels:
            kubernetes.io/metadata.name: monitoring
# ...
----
====

> You can use an `AdminNetworkPolicy` resource with a `Pass` value for the `action` field in conjunction with the `BaselineAdminNetworkPolicy` resource to create a multi-tenant policy that allows one tenant to collect monitoring data on their application while simultaneously not collecting data from a second tenant.

Reviewer 2: Long sentence. Up to you whether you fix it. It's 34 words. 

```suggestion
You can use an `AdminNetworkPolicy` resource with a `Pass` value for the `action` field in conjunction with the `BaselineAdminNetworkPolicy` resource to create a multi-tenant policy. This multi-tenant policy allows one tenant to collect monitoring data on their application while simultaneously not collecting data from a second tenant.
```

> As an administrator if you apply both the "AdminNetworkPolicy `Pass` action example" and the "BaselineAdminNetwork Policy `Deny` example", tenants are then left with the ability to choose to create a `NetworkPolicy` resource that will be evaluated before the BANP.

Reviewer 2: ```suggestion
As an administrator, if you apply both the "AdminNetworkPolicy `Pass` action example" and the "BaselineAdminNetwork Policy `Deny` example", tenants are then left with the ability to choose to create a `NetworkPolicy` resource that will be evaluated before the BANP.
```

For example, Tenant 1 can set up the following `NetworkPolicy` resource to monitor ingress traffic:

.Example `NetworkPolicy`
[%collapsible]
====
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-monitoring
  namespace: tenant 1
spec:
  podSelector:
  policyTypes:
    - Ingress
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: monitoring
# ...
----
====

> In this scenario, tenant 1's policy would be evaluated after the "AdminNetworkPolicy `Pass` action example" and before the "BaselineAdminNetwork Policy `Deny` example", which denies all ingress monitoring traffic coming into tenants with `security` level `internal`. With Tenant 1's `NetworkPolicy` object in place, they will be able to collect data on their application while tenant 2 who does not have any `NetworkPolicy` objects in place will not be able to collect data, and as an administrator, you have not by default monitored internal tenants. Instead, you created a BANP that allows tenants to use `NetworkPolicy` objects to override the default behavior of your BANP.

Reviewer 2: Tenant 1 is capitalized above but lower case here. 

tenant 2 is not capitalized. 

Does security need to be in backticks here? 

The second sentence probably needs broken up too. I'm having a problem rewriting the third part. so maybe consider: 

```suggestion
In this scenario, Tenant 1's policy would be evaluated after the "AdminNetworkPolicy `Pass` action example" and before the "BaselineAdminNetwork Policy `Deny` example", which denies all ingress monitoring traffic coming into tenants with `security` level `internal`. With Tenant 1's `NetworkPolicy` object in place, they will be able to collect data on their application. Meanwhile, Tenant 2, who does not have any `NetworkPolicy` objects in place, will not be able to collect data. As an administrator, you have not by default monitored internal tenants, but instead, you created a BANP that allows tenants to use `NetworkPolicy` objects to override the default behavior of your BANP.
```

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: Concept
[id="BaselineAdminNetworkPolicy"_{context}"]
= BaselineAdminNetworkPolicy

> `BaselineAdminNetworkPolicy` (BANP) is a cluster-scoped custom resource definitions (CRDs). As a {product-title} administrator, you can use BANP to setup and enforce optional baseline network policies that are overridable by users using `NetworkPolicies` if need be. Rule actions for BANP are `allow` or `deny`.

Reviewer: Same changes as requested for ANP: https://github.com/openshift/openshift-docs/pull/70800#discussion_r1482995038
```suggestion
`BaselineAdminNetworkPolicy` (BANP) is a cluster-scoped custom resource definition (CRD). As a {product-title} administrator, you can use BANP to setup and enforce optional baseline network policie rules that are overridable by users using `NetworkPolicies` if need be. Rule actions for BANP are `allow` or `deny`.
```
Add https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy#story-5-cluster-wide-default-guardrails to this aspect so that users know why BANP is present in addition to ANP.
there are two technical aspects here:

1. it can act as the catch-all for passed traffic
2. it can act as the default cluster security guardrail

Also worth mentioning at a given time there can only be 1 BANP in the cluster. It is a singleton object.

Author: This information is below. I'll move it up.

* A `subject` that consists of a set of pods that is selected by the policy.

* A list of ingress rules to be applied for all ingress traffic towards the `subject`.

* A list of egress rules to be applied for all egress traffic from the `subject`.

[NOTE]
====
`BaselineAdminNetworkPolicy` is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`.
====

[discrete]
[id="BaselineAdminNetworkPolicy-example"]
== BaselineAdminNetworkPolicy example

The `BaselineAdminNetworkPolicy` resource is a cluster singleton `object` that can be used as a guardrail policy incase a passed traffic policy does not match any `NetworkPolicies` in the cluster. BANP can also be used as a default security model that provides guardrails that intra-cluster traffic is blocked by default and user will need to use `NetworkPolicy` to allow known traffic. You must use `default` as the name when creating a `BaselineAdminNetworkPolicy` object.

.Example YAML file for BANP
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:

>   name: BANP_example <1>

Reviewer: this is wrong, you can't have any other name other than `default` which brings me to my next question @JoeAldinger have you tested all the yamls on a live cluster? Users won't like it if they copy this and it doesn't work.

Reviewer: also just noticed this is still a ANP example?

Reviewer: and there is no pass in BANP, the rule name suggests otherwise

spec:
  subject:
    namespaces:
      matchLabels:

>           kubernetes.io/metadata.name: gryffindor

Reviewer 3: Probably need to swap out all the "gryffindor", "slytherin", and "conformance-house" example values/fields throughout to something more generic.

>   ingress:<2>

Reviewer 3: ```suggestion
  ingress: <2>
```

  - name: "deny-all-ingress-from-slytherin"
    action: "Deny"
    from:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              conformance-house: slytherin
        podSelector:
          matchLabels:
            conformance-house: slytherin

>   egress:<2>

Reviewer 3: ```suggestion
  egress: <2>
```

  - name: "pass-all-egress-to-slytherin"
    action: "Allow" <3>
    to:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              conformance-house: slytherin
        podSelector:
          matchLabels:
            conformance-house: slytherin
----
<1> Policy name
<2> BANP have both ingress and egress rules.

> <3> BANP `action` rules for ingress and egress only accepts values of `Deny` and `Allow`.

Reviewer 3: Need to sync up one way or the other on punctuation or no-punctuation between these 3.

====

[discrete]
[id="BaselineAdminNetworkPolicy-default-deny-example"_{context}]
== BaselineAdminNetworkPolicy Deny example

> The following BANP singleton ensures that the administrator has set up a default deny policy for all ingress monitoring traffic coming into the tenants at `internal` security level. When combined with the "AdminNetworkPolicy Pass example", this deny policy acts as a guardrail policy for all ingress traffic that is passed by the ANP `pass-monitoring` policy.

Reviewer 3: I don't have a suggestion really, but if there's another word that could be used in place of "guardrail", might be worth switching.

Edit: Seems that that would involve changing quite a few instances, so probably fine as-is, but something to ponder.

Author: Yeah, I don't disagree. I just couldn't think of suitable alternative. I'll keep thinking about it though.

> .Example YAML file for a guardrail Deny rule

Reviewer 3: ```suggestion
.Example YAML file for a guardrail `Deny` rule
```

[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces:
      matchLabels:
        security: internal
  ingress:
  - name: "deny-ingress-from-monitoring"
    action: "Deny"
    from:
    - namespaces:
        namespaceSelector:
          matchLabels:
# ...
----
====

> //peer-reviewer, please stop here in this module. I will be pulling this out and placing it in a module for 4.16. Time crunch, so I wanted to get peer-review rolling on this.

Reviewer 3: Stopped here in this module per comment.

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: Concept
[id="BaselineAdminNetworkPolicy"_{context}"]
= BaselineAdminNetworkPolicy

`BaselineAdminNetworkPolicy` (BANP) is a cluster-scoped custom resource definition (CRD). As a {product-title} administrator, you can use BANP to setup and enforce optional baseline network policy rules that are overridable by users using `NetworkPolicy` objects if need be. Rule actions for BANP are `allow` or `deny`.

The `BaselineAdminNetworkPolicy` resource is a cluster singleton object that can be used as a guardrail policy incase a passed traffic policy does not match any `NetworkPolicy` objects in the cluster. A BANP can also be used as a default security model that provides guardrails that intra-cluster traffic is blocked by default and a user will need to use `NetworkPolicy` objects to allow known traffic. You must use `default` as the name when creating a BANP resource.

A BANP allows administrators to specify:

> * A `subject` that consists of a set of pods that is selected by the policy.

Reviewer 4: subject is set of namespaces or namespace, selection is not at pod level.

.Example YAML file for BANP
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default <1>
spec:
  subject:
    namespaces:
      matchLabels:
          kubernetes.io/metadata.name: example.name <2>
  ingress: <3>
  - name: "deny-all-ingress-from-tenant-1" <4>
    action: "Deny"
    from:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              custom-banp: tenant-1 <5>
        podSelector:
          matchLabels:
            custom-banp: tenant-1 <6>
  egress:
  - name: "allow-all-egress-to-tenant-1"
    action: "Allow"
    to:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              custom-banp: tenant-1
        podSelector:
          matchLabels:
            custom-banp: tenant-1
----
<1> The policy name must be `default` because BANP is a singleton object.
<2> Specify the namespace to apply the ANP to.
<3> BANP have both ingress and egress rules. BANP rules for `spec.ingress` and `spec.egress` fields accepts values of `Deny` and `Allow` for the `action` field.
<4> Specify a name for the `ingress.name`

> <5> Specify the pods in the namespaces to apply the BANP resource.

Reviewer 4: This should be reworded to say specify the namespaces to apply the BANP resource.
podSelector is to narrow down the pods in selected namespaces.

//[discrete]
//[id="BaselineAdminNetworkPolicy-multi-tenancy-example"_{context}]
//== BaselineAdminNetworkPolicy multi-tenancy example

The following example demonstrates how a `AdminNetworkPolicy` resource using a `Pass` value for the `action` field can be used by administrators in conjunction with the `BaselineAdminNetworkPolicy` resource to create a multi-tenancy policy that allows one tenant to collect monitoring data on their application while simultaneously not collecting data from a second tenant.

As an administrator if you apply both the "AdminNetworkPolicy `Pass` action example" and the "BaselineAdminNetwork Policy `Deny` example", tenants are then left with the ability to choose to create a `NetworkPolicy` policy that will be evaluated before the BANP.

For example, Tenant 1 can set up the following `NetworkPolicy` policy to monitor ingress traffic:

.Example `NetworkPolicy`
[%collapsible]
====
[source,yaml]
----
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-monitoring
  namespace: tenant 1
spec:
  podSelector:
  policyTypes:
    - Ingress
  ingress:
  - from:
    - namespaceSelector:

>         matchLabels:

Reviewer 4: <pre>
kubernetes.io/metadata.name: monitoring
</pre>
